lib/cartopy/mpl/gridliner.py

```diff
@@ -327,58 +338,257 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
+                            # and they intersect themselve too musc
+                            # it results in a multiline string
+                            # that must be converted to a single linestring
+                            x, y = intersection[0].coords.xy
+                            x += intersection[-1].coords.xy[0]
+                            y += intersection[-1].coords.xy[1]
+                            xy = np.array((x, y))
+                            intersection = [sgeom.LineString(xy.T)]
+                        tails = []
+                        heads = []
+                        for inter in intersection:
+                            if len(inter.coords) < 2:
+                                continue
+                            tails.append(inter.coords[:2])
+                            heads.append(inter.coords[-1:-3:-1])
+                        if not tails:
+                            continue
+                    else:
+                        # TODO: we should handle GeometryCollection
+                        continue
+                    del intersection
+
+                    # Loop on head and tail and plot label by extrapolation
+                    for tail, head in zip(tails, heads):
+                        for pt0, pt1 in [tail, head]:
+                            kw, angle, loc = self._segment_to_text_specs(
+                                    pt0, pt1)
+                            if not getattr(self, loc+'_labels'):
+                                continue
+                            kw.update(label_style,
+                                      bbox={'pad': 0, 'visible': False})
+                            x = round(pt0[0], 5)
+                            y = round(pt0[1], 5)
+                            text = formatter(tick_value)
+                            tt = self.axes.text(pt0[0], pt0[1], text, **kw)
+                            tt._angle = angle
+                            priority = (((lonlat == 'lon') and
+                                         loc in ('bottom', 'top')) or
+                                        ((lonlat == 'lat') and
+                                         loc in ('left', 'right')))
+                            self._labels.append((priority, tt))
+                            getattr(self, loc + '_label_artists').append(tt)
+
+        # Sort labels
+        if self._labels:
+            self._labels.sort(key=operator.itemgetter(0), reverse=True)
+            self._update_labels_visibility(renderer)
+
+    def _segment_to_text_specs(self, pt0, pt1):
+        """Get appropriate kwargs for a label from lon or lat line segment"""
+        x0, y0 = pt0
+        x1, y1 = pt1
+        angle = np.arctan2(y0-y1, x0-x1) * 180 / np.pi
+        kw, loc = self._segment_angle_to_text_specs(angle)
+        return kw, angle, loc
+
+    def _segment_angle_to_text_specs(self, angle):
+        """Get appropriate kwargs for a given direction angle"""
```

Comment:
I think that `Get appropriate kwargs for a given text angle` is clearer.

Reply:
Fixed